ouroboros-consensus-test/test-consensus/Test/Consensus/Mempool/Fairness.hs

```diff
@@ -0,0 +1,207 @@
+                  ]
+
+type TestMempool = Mempool IO TestBlock
+
+-- | Test if the mempool treats small and large transactions in the same way.
+--
+-- We run the following test:
+--
+-- - Given a mempool 'mp'.
+-- - In parallel:
+--     - Run 'N' threads that add small transactions to 'mp'.
+--     - Run 'N' threads that add large transactions to 'mp'.
+--     - Remove transactions from 'mp' one by one, with a small delay between
+--       removals. Collect the removed transactions.
+--
+-- After collecting 'M' removed transactions, let 'diff' be the difference between
+-- the number of small and large transactions that were added to 'mp', then we
+-- check that:
+--
+-- > diff / M <= toleranceThreshold
+--
+-- See 'TestParams' for an explanation of the different parameters that
+-- influence this test.
+testTxSizeFairness :: TestParams -> IO ()
+testTxSizeFairness TestParams { mempoolMaxCapacity, smallTxSize, largeTxSize, nrOftxsToCollect, toleranceThreshold } = do
+    ----------------------------------------------------------------------------
+    --  Obtain a mempool.
+    ----------------------------------------------------------------------------
+    let
+      ledgerItf = Mempool.LedgerInterface {
+              Mempool.getCurrentLedgerState = pure $ testInitLedgerWithState ()
+          }
+
+      sampleLedgerConfig =
+          HardFork.defaultEraParams (Consensus.SecurityParam 10) (Time.slotLengthFromSec 2)
+    mempool <- Mempool.openMempoolWithoutSyncThread
+                   ledgerItf
+                   sampleLedgerConfig
+                   (Mempool.mkCapacityBytesOverride mempoolMaxCapacity)
+                   Tracer.nullTracer
+                   genTxSize
+
+    ----------------------------------------------------------------------------
+    --  Add and collect transactions
+    ----------------------------------------------------------------------------
+    let runConcurrently = Async.runConcurrently . asum . fmap Async.Concurrently
+    txs <- runConcurrently [                      adders  mempool smallTxSize
+                           , threadDelay 1_000 >> adders  mempool largeTxSize
+                           , threadDelay 1_000 >> remover mempool             nrOftxsToCollect
+                           ]
+    -- We want to give the small adders the head start, to make sure the mempool
+    -- is filled up with small transactions first. That is why we use
+    -- threadDelay above.
+
+    ----------------------------------------------------------------------------
+    --  Count the small and large transactions
+    ----------------------------------------------------------------------------
+    let
+      nrSmall :: Double
+      nrLarge :: Double
+      (nrSmall, nrLarge) = (fromIntegral . length *** fromIntegral . length)
+                         $ List.partition  (<= smallTxSize)
+                         $ fmap txSize txs
+    length txs @?= nrOftxsToCollect
```

Comment:
Suggestion: seems like a different assertion? Perhaps make it clearer what we are asserting:
```
length txs @?= nrOftxsToCollect
largeOverSmallRatioBelowThreshold txs toleranceThreshold
```

Reply:
Define largeOverSmallRatioBelowThreshold locally.